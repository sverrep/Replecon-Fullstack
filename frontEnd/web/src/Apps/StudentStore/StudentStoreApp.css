.store-content{
    margin: 0 auto;
    width: 1080px;
}

.item-cards{
    margin-top: 25px;
    margin-left: 0px;
}

.cards{
    margin-right: 20px;
}

.title{
    display: flex;
    flex-direction: row;
    margin-top: 20px;
}

.title-name{
    text-align: left;
}

.title-balance{
    float: right;
    margin-left:auto;
}


.wrapper{
    font-family: 'Montserrat', sans-serif;
}


.alert{
    margin-left: 200px;
    margin-right: 200px;
}

.store-button{
    background-color: #546F99 !important;
    border-color: #546F99 !important;
}

