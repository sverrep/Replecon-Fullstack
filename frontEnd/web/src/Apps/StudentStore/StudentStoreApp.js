import React from 'react';
import { withRouter } from "react-router-dom";
import axios from 'axios';
import getIP from '../../settings.js';
import NavBar from '../../Components/navbar/Navbar.js';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import Row from 'react-bootstrap/Row';
import './StudentStoreApp.css';
class StudentStore extends React.Component {
    constructor(props) {
        
        super(props);
        this.state = {
            student_balance: '',
            class_code:'',
            store_name:'',
            store_id: 0,
            shops:[],
            students: [],
            items: [],

        }
        this.getStudentBalance = this.getStudentBalance.bind(this)
        this.getClassCode = this.getClassCode.bind(this)
    }

    componentDidMount(){
        this.getStudentBalance()
        this.getClassCode()
    }

    //Classroom Store Setups
    getClassCode(){
        axios.get(getIP()+'/students/class_code/')
        .then(response => {
        this.setState({ students: response.data });
        this.setState({ class_code: this.state.students[0].class_code })
        this.getShops()
        })
        .catch(error => console.log(error))
    }

    getShops(){
        axios.get(getIP()+'/shops/')
        .then(response => {
        this.setState({shops: response.data})
        this.findShopName(response.data)
        })
        .catch(error => console.log(error))
    }

    findShopName(shops){
        for (let i = 0; i<=shops.length -1;i++)
        {
        if(shops[i].class_code===this.state.class_code)
        {
            this.setState({store_name:shops[i].shop_name})
            this.setState({store_id: shops[i].id})

        }
        }
        this.getItems()
    }

    getItems(){
        axios.get(getIP()+'/items/')
        .then(response => {
        this.getShopItems(response.data)
        })
        .catch(error => console.log(error))

    }

    getShopItems(allItems){
        var ar = []
        for(let i = 0; i<=allItems.length -1; i++)
        {
        if(allItems[i].shop_id===this.state.store_id)
        {
            ar.push(allItems[i])
        }
        }
        this.setState({items:ar})
        }

    //Student Balance
    getStudentBalance()
    {
        axios.get(getIP()+'/students/balance/')
        .then(response => {
            this.setState({student_balance: response.data})
            })
        .catch(error => console.log(error))
    }

    renderCard(item){
        return(
            <Card style={{ width: '15rem', padding: '10px' }} className='cards'>
                <Card.Body>
                    <Card.Title>{item.item_name}</Card.Title>
                    <Card.Text>
                        {item.description}
                    </Card.Text>
                    <Button variant="primary" onClick={() => this.buyItem(item)}>Purchase</Button>
                </Card.Body>
            </Card>
        )
    }


    //Purchasing of Item

    buyItem(item){
        axios.post(getIP()+'/items/boughtitems/', { item_name: item.item_name })
            .then(response => {
                axios.get(getIP()+'/students/store/')
                    .then(response => {
                    axios.put(getIP()+'/students/balance/', { amount: item.price, user_id: response.data, recipient: false })
                        .then(response => {
                        axios.post(getIP()+'/transactions/buyFromStore/', { amount: item.price })
                            .then(response => {
                                this.getStudentBalance()
                            })
                        .catch(error => console.log(error + "transactions"))
                    })
                    . catch(error => console.log(error + "students"))
                })
                .catch(error => console.log(error + "store account"))
            })
            .catch(error => console.log(error + "items"))
            
        
    }
    render(){
        return(
            <div className='wrapper'>
                <NavBar/>
                <div className='title'>
                    <h3>{this.state.store_name}</h3>
                    <h5>Your current balance: {this.state.student_balance} </h5>
                </div>
                <div className="item-cards">
                    <div className='override'>
                        <Row xs="auto" md={2}>
                    {this.state.items.map(item => {
                            return(this.renderCard(item))
                        })}
                        </Row>
                    </div>
                </div>
            </div>

        )
    }

}
export default withRouter(StudentStore)