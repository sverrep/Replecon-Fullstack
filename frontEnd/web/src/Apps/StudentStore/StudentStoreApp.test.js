import React from 'react';
import ReactDom from 'react-dom';
import StudentStoreApp from './StudentStoreApp.js'
import renderer from 'react-test-renderer'
import axios from 'axios'

jest.mock('axios');
let wrapper
const props = {
    location:{
        state:{
            role: "Student",
            token: 'Token'
        }
    }  
}


describe('Student Store Tests', ()=>{
    
    beforeEach(() => {
        wrapper = shallow(<StudentStoreApp.WrappedComponent {...props}/>, { disableLifecycleMethods: true })
    });
    
    describe('Student Page renders', ()=>{
        it('should render page properly',()=>{
            expect(wrapper.getElements()).toMatchSnapshot();
        })

        it('should render cards properly', ()=>{
            const instance = wrapper.instance()
            expect(instance.renderCard({shop_id: '12', item_name: 'item name', price: '60', description: 'description'})).toMatchSnapshot();
        })
    })

    describe('Component did mount function', ()=>{
        it('component did mount should call other functions', async()=>{
            const instance = wrapper.instance()
            
            instance.getClassCode = jest.fn()
            instance.getClassCode.mockReturnValueOnce()

            instance.getStudentBalance = jest.fn()
            instance.getStudentBalance.mockReturnValueOnce()
            
            await instance.componentDidMount()

            expect(instance.getClassCode).toHaveBeenCalledTimes(1)
            expect(instance.getStudentBalance).toHaveBeenCalledTimes(1)
        })
    })

    describe('Component Mounted Unit Tests', ()=>{
        it('Should get class codes', async()=>{
            const instance = wrapper.instance()
            
            const users = mockResponseData([{id: '1', class_code: '123'}])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(users))
            
            instance.getShops = jest.fn()
            instance.getShops.mockReturnValueOnce()

            await instance.getClassCode()
            expect(instance.state.students).toStrictEqual([{id: '1', class_code: '123'}])
            expect(instance.state.class_code).toBe('123')
            expect(instance.getShops).toHaveBeenCalledTimes(1)
        })

        it('should get shops', async()=>{
            const instance = wrapper.instance()

            const shops = mockResponseData([{class_code: '123', shop_name: 'shop name', id: '12'}, {class_code: '34', shop_name: 'shop name2', id: '122'}])
            axios.get.mockImplementationOnce(() =>  Promise.resolve(shops))

            instance.findShopName = jest.fn()
            instance.findShopName.mockReturnValueOnce()

            await instance.getShops()

            expect(instance.state.shops).toStrictEqual([{class_code: '123', shop_name: 'shop name', id: '12'}, {class_code: '34', shop_name: 'shop name2', id: '122'}])
            expect(instance.findShopName).toHaveBeenCalledTimes(1)
        })

        it('should find shop name', async()=>{
            const instance = wrapper.instance()
            instance.state.class_code = '34'
            const shops = [{class_code: '123', shop_name: 'shop name', id: '12'}, {class_code: '34', shop_name: 'shop name2', id: '122'}]
            
            instance.getItems = jest.fn()
            instance.getItems.mockReturnValueOnce()

            instance.findShopName(shops)
            
            expect(instance.state.store_name).toBe('shop name2')
            expect(instance.state.store_id).toBe('122')
        })

        it('it should get all items', async()=>{
            const instance = wrapper.instance()
            instance.state.store_id = '12'
            const allitems = [{shop_id: '12', item_name: 'item name', item_price: '60'}, {shop_id: '15', item_name: 'item name3', item_price: '20'}]

            instance.getShopItems(allitems)
            expect(instance.state.items).toStrictEqual([{shop_id: '12', item_name: 'item name', item_price: '60'}])
        })
    })

    describe('Component Mounted Functions Integration', ()=>{

        
        it('should get student balance', async()=>{           
            const instance = wrapper.instance()
            const bal = mockResponseData('100')
            await axios.get.mockImplementationOnce(() =>  Promise.resolve(bal))
            await instance.getStudentBalance()
            
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(instance.state.student_balance).toBe('100')
        })

        it('should get student class code -> shop -> items', async()=>{
            const instance = wrapper.instance()
            
            const users = mockResponseData([{id: '1', class_code: '123'}])
            await axios.get.mockImplementationOnce(() =>  Promise.resolve(users))

            const shops = mockResponseData([{class_code: '123', shop_name: 'shop name', id: '12'}, {class_code: '34', shop_name: 'shop name2', id: '122'}])
            await axios.get.mockImplementationOnce(() =>  Promise.resolve(shops))

            const items = mockResponseData([{shop_id: '12', item_name: 'item name', item_price: '60'}])
            await axios.get.mockImplementationOnce(() =>  Promise.resolve(items))

            await instance.getClassCode()
            
            
            expect(axios.get).toHaveBeenCalledTimes(3)
            expect(instance.state.class_code).toBe('123')
            expect(instance.state.shops).toStrictEqual([{class_code: '123', shop_name: 'shop name', id: '12'}, {class_code: '34', shop_name: 'shop name2', id: '122'}])
            expect(instance.state.store_name).toBe('shop name')
            expect(instance.state.items).toStrictEqual([{shop_id: '12', item_name: 'item name', item_price: '60'}])
        })
        

    })

    describe('Purchasing', ()=>{
        it('should pass purchase validation with enough money', ()=>{
            const instance = wrapper.instance()
            const item = {shop_id: '12', item_name: 'item name', price: '60', description: 'description'}
            instance.state.student_balance = 100
            const result = instance.validatePurchase(item)
            expect(result).toBe(true)

        })

        it('should fail purchase validation with not enough money',()=>{
            const instance = wrapper.instance()
            const item = {shop_id: '12', item_name: 'item name', price: '60', description: 'description'}
            instance.state.student_balance = 40
            const result = instance.validatePurchase(item)
            expect(result).toBe(false)
        })

        it('should buy item for student', async()=>{
            const instance = wrapper.instance()
            const item = {id: '1', shop_id: '12', item_name: 'item name', price: '60', description: 'description'}
            instance.validatePurchase = jest.fn()
            instance.validatePurchase.mockReturnValueOnce(true)
            
            const payload = {item_id: item.id}
            axios.post.mockImplementationOnce(() => Promise.resolve())
            
            const user_id = mockResponseData('2')
            axios.get.mockImplementationOnce(() =>  Promise.resolve(user_id))
            
            const payload2 = { amount: item.price, user_id: '2', recipient: false }
            axios.put.mockImplementationOnce(() =>  Promise.resolve())

            const payload3 = {amount: item.price}
            axios.post.mockImplementationOnce(() => Promise.resolve())

            instance.getStudentBalance = jest.fn()
            instance.getStudentBalance.mockReturnValueOnce()

            await instance.buyItem(item)
            expect(axios.get).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledTimes(2)
            expect(axios.put).toHaveBeenCalledTimes(1)
            expect(axios.post).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//items/boughtitems/', payload)
            expect(axios.get).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//students/store/')
            expect(axios.put).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//students/balance/', payload2)
            expect(axios.post).toHaveBeenCalledWith("https://mythical-mason-324813.ey.r.appspot.com//transactions/buyFromStore/", payload3)
            expect(instance.getStudentBalance).toHaveBeenCalledTimes(1)
        })
    })

})


const mockResponseData = jest.fn((data) => {
    if (typeof data === 'string')
    {
        return {
            data: data,
        };
    }
    else
    {
        return {
            data: data,
        };
    }
  });