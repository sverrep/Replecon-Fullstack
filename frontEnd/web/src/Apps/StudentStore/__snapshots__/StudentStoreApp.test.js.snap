// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Store Tests Student Page renders should render cards properly 1`] = `
<Card
  body={false}
  className="cards"
  style={
    Object {
      "padding": "10px",
      "width": "15rem",
    }
  }
>
  <CardBody>
    <CardTitle>
      item name
       $
      60
    </CardTitle>
    <CardText>
      description
    </CardText>
    <Button
      active={false}
      className="store-button"
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Purchase
    </Button>
  </CardBody>
</Card>
`;

exports[`Student Store Tests Student Page renders should render page properly 1`] = `
Array [
  <div
    className="wrapper"
  >
    <Navbar
      className="teacher-navbar"
      collapseOnSelect={true}
      expand="lg"
      variant="dark"
    >
      <Container
        fluid={false}
      >
        <NavbarBrand
          className="navbar-header"
        >
          Replecon
        </NavbarBrand>
        <Nav
          className="navbar-btns"
          fill={false}
          justify={false}
        >
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/Profile",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Profile
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentStore",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Store
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentBank",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Bank
            </Link>
          </Button>
          <Button
            active={false}
            className="navbar-btn"
            disabled={false}
            variant="light"
          >
            <Link
              className="navbar-link"
              to={
                Object {
                  "pathname": "/StudentClass",
                  "state": Object {
                    "role": "Student",
                    "token": "Token",
                  },
                }
              }
            >
              Classroom
            </Link>
          </Button>
        </Nav>
      </Container>
    </Navbar>
    <div
      className="store-content"
    >
      <div
        className="title"
      >
        <div
          className="title-name"
        >
          <h3>
            
          </h3>
        </div>
        <div
          className="title-balance"
        >
          <h3>
            $
            
             
          </h3>
        </div>
      </div>
      <div
        className="alert"
      />
      <div
        className="item-cards"
      >
        <div
          className="override"
        >
          <Row
            md={2}
            xs="auto"
          />
        </div>
      </div>
    </div>
  </div>,
]
`;
